src/connection/host_lookup.c

```diff
@@ -0,0 +1,444 @@
+ *
+ */
+
+/*
+ * host_lookup.c
+ */
+
+#ident "$Id$"
+
+#include <stdio.h>
+#include <string.h>
+#include <limits.h>
+#include <errno.h>
+#include <unordered_map>
+
+#if defined (WINDOWS)
+#include <winsock2.h>
+#include <ws2tcpip.h>
+#else
+#include <arpa/inet.h>
+#include <netdb.h>
+#endif
+
+#include "system_parameter.h"
+#include "environment_variable.h"
+
+#define HOSTNAME_BUF_SIZE              (256)
+#define MAX_HOSTS_LINE_NUM       (256)
+#define IPADDR_LEN              (17)
+
+#define INSERT_IPADDR 0x0
+#define INSERT_HOSTNAME 0x1
+
+typedef enum
+{
+  HOSTNAME_TO_IPADDR = 0,
+  IPADDR_TO_HOSTNAME = 1,
+} LOOKUP_TYPE;
+
+typedef enum
+{
+  USE_GLIBC_HOSTS = 0,
+  USE_USER_DEFINED_HOSTS = 1,
+} HOST_LOOKUP_TYPE;
+
+struct cub_hostent
+{
+  char ipaddr[IPADDR_LEN];
+  char hostname[HOSTNAME_BUF_SIZE + 1];
+};
+
+typedef struct cub_hostent CUB_HOSTENT;
+static CUB_HOSTENT hostent_List[MAX_HOSTS_LINE_NUM];	/* "etc/hosts" file maximum line */
+//should I need to allocate all 256 element??
+static const char user_defined_hostfile_Name[] = "hosts.conf";
+
+static int host_conf_element_Count = 0;
+
+// *INDENT-OFF*
+static std::unordered_map <std::string, std::string> user_host_Map;
+// *INDENT-ON*
+
+static int host_conf_load ();
+static struct hostent *host_lookup_internal (const char *hostname, struct sockaddr *saddr, LOOKUP_TYPE lookup_case);
+
+static struct hostent *
+host_lookup_internal (const char *hostname, struct sockaddr *saddr, LOOKUP_TYPE lookup_case)
+{
+  static struct hostent hp;
+  int i, find_index = -1;
+
+  char addr_transform_buf[IPADDR_LEN];
+  char addr_transform_buf2[IPADDR_LEN];
+  struct sockaddr_in *addr_trans = NULL;
+
+  if (host_conf_element_Count == 0)
+    {
+      host_conf_element_Count = host_conf_load ();
+      if (host_conf_element_Count <= 0)
+	{
+//err_msg : load fail
+	  return NULL;
+	}
+    }
+
+  addr_trans = (struct sockaddr_in *) saddr;
+
+  assert (((hostname != NULL) && (saddr == NULL)) || ((hostname == NULL) && (saddr != NULL)));
+
+  if (lookup_case == IPADDR_TO_HOSTNAME)
+    {
+      if (inet_ntop (AF_INET, &addr_trans->sin_addr, addr_transform_buf, sizeof (addr_transform_buf)) == NULL)
+	{
+//err_msg : convert binary to text fail
+	  return NULL;
+	}
+
+    }
+
+/*Look up in the hostent_List*/
+  for (i = 0; i < host_conf_element_Count; i++)
+    {
+      if ((lookup_case == HOSTNAME_TO_IPADDR) && (strcmp (hostname, hostent_List[i].hostname) == 0))
+	{
+	  find_index = i;
+	  break;
+	}
+
+      if ((lookup_case == IPADDR_TO_HOSTNAME) && (strcmp (addr_transform_buf, hostent_List[i].ipaddr) == 0))
+	{
+	  find_index = i;
+	  break;
+	}
+    }
+
+  if (find_index < 0)
+    {
+      return NULL;
+//err_msg : hosts.conf has no info
+    }
+
+  if (inet_pton (AF_INET, hostent_List[find_index].ipaddr, addr_transform_buf2) < 1)
+    {
+//err_msg : convert binary to text fail -> reverse
+      return NULL;
+    }
+
+  /*set the hostent struct for the return value */
```

Comment:
How about make the new function to set hp value?

Reply:
And malloc error must be added.

Reply:
`free(NULL)` is not invalid, so I will use it to free memory allocation together.